File: src/zdecode/zdecode.py
#!/usr/bin/env python3
# Cyborg ZOSCII v20260301
# (c) 2026 Cyborg Unicorn Pty Ltd.
# This software is released under MIT License.

import sys
import struct
import os

class RomData:
    def __init__(self):
        self.ptrROMData = b''
        self.lngROMSize = 0

ZOSCII_ROM_LOAD_MAX = 131072

def loadRom(strFilename_a):
    ptrRom = None
    
    if os.path.exists(strFilename_a):
        try:
            with open(strFilename_a, 'rb') as ptrFile:
                arrBuf = ptrFile.read(ZOSCII_ROM_LOAD_MAX)
                if arrBuf:
                    ptrRom = RomData()
                    ptrRom.ptrROMData = arrBuf
                    ptrRom.lngROMSize = len(arrBuf)
        except IOError:
            ptrRom = None
    
    return ptrRom

def unloadRom(ptrRom_a):
    # In Python, garbage collector handles this, but method kept for symmetry
    ptrRom_a.ptrROMData = b''
    ptrRom_a.lngROMSize = 0

def decodeFile(ptrRom_a, strInputFile_a, strOutputFile_a):
    blnSuccess = False
    ptrInput = None
    ptrOutput = None
    
    if os.path.exists(strInputFile_a):
        try:
            ptrInput = open(strInputFile_a, 'rb')
            
            # Get input file size
            ptrInput.seek(0, 2)
            lngInputSize = ptrInput.tell()
            ptrInput.seek(0, 0)
            
            lngSlots = lngInputSize // 2
            
            if lngSlots >= 0:
                ptrOutput = open(strOutputFile_a, 'wb')
                
                # Decode each slot
                for intI in range(lngSlots):
                    arrBuf = ptrInput.read(2)
                    if len(arrBuf) != 2:
                        break
                    intAddr = struct.unpack('<H', arrBuf)[0]
                    if intAddr < ptrRom_a.lngROMSize:
                        ptrOutput.write(bytes([ptrRom_a.ptrROMData[intAddr]]))
                
                if lngSlots > 0 and intI == lngSlots - 1:
                    blnSuccess = True
                
                ptrOutput.close()
            ptrInput.close()
        except IOError:
            if ptrOutput:
                ptrOutput.close()
            if ptrInput:
                ptrInput.close()
    
    return blnSuccess

def main():
    intResult = 1
    ptrRom = None
    blnDecodeOk = False
    
    print("ZOSCII Decoder")
    print("(c) 2026 Cyborg Unicorn Pty Ltd v20260301 - MIT License\n")
    
    if len(sys.argv) == 4:
        ptrRom = loadRom(sys.argv[1])
        if ptrRom:
            blnDecodeOk = decodeFile(ptrRom, sys.argv[2], sys.argv[3])
            
            if blnDecodeOk:
                intResult = 0
                print("Decode successful!")
            else:
                print("Decode failed", file=sys.stderr)
            
            unloadRom(ptrRom)
        else:
            print("Failed to load ROM", file=sys.stderr)
    else:
        print(f"Usage: {sys.argv[0]} <romfile> <encoded> <output>", file=sys.stderr)
    
    sys.exit(intResult)

if __name__ == "__main__":
    main()

File: src/zdecode/zdecode.php
<?php
// Cyborg ZOSCII v20260301
// (c) 2026 Cyborg Unicorn Pty Ltd.
// This software is released under MIT License.

class RomData 
{
    public $ptrROMData = '';
    public $lngROMSize = 0;
}

define('ZOSCII_ROM_LOAD_MAX', 131072);

function loadRom($strFilename_a) 
{
    $ptrRom = null;
    
    if (file_exists($strFilename_a)) 
    {
        $ptrRom = new RomData();
        $ptrRom->ptrROMData = file_get_contents($strFilename_a);
        if ($ptrRom->ptrROMData !== false) 
        {
            $ptrRom->lngROMSize = strlen($ptrRom->ptrROMData);
            if ($ptrRom->lngROMSize > ZOSCII_ROM_LOAD_MAX) 
            {
                $ptrRom->ptrROMData = substr($ptrRom->ptrROMData, 0, ZOSCII_ROM_LOAD_MAX);
                $ptrRom->lngROMSize = ZOSCII_ROM_LOAD_MAX;
            }
        } 
        else 
        {
            $ptrRom = null;
        }
    }
    
    return $ptrRom;
}

function unloadRom($ptrRom_a) 
{
    // In PHP, garbage collector handles this, but method kept for symmetry
    $ptrRom_a->ptrROMData = '';
    $ptrRom_a->lngROMSize = 0;
}

function decodeFile($ptrRom_a, $strInputFile_a, $strOutputFile_a) 
{
    $blnSuccess = false;
    $ptrInput = null;
    $ptrOutput = null;
    $arrBuf = '';
    $lngInputSize = 0;
    $lngSlots = 0;
    $intI = 0;
    
    if (file_exists($strInputFile_a)) 
    {
        $ptrInput = fopen($strInputFile_a, 'rb');
        if ($ptrInput) 
        {
            // Get input file size
            fseek($ptrInput, 0, SEEK_END);
            $lngInputSize = ftell($ptrInput);
            fseek($ptrInput, 0, SEEK_SET);
            
            $lngSlots = $lngInputSize / 2;
            
            if ($lngSlots >= 0) 
            {
                $ptrOutput = fopen($strOutputFile_a, 'wb');
                if ($ptrOutput) 
                {
                    // Decode each slot
                    for ($intI = 0; $intI < $lngSlots; $intI++) 
                    {
                        $arrBuf = fread($ptrInput, 2);
                        if (strlen($arrBuf) != 2) 
                        {
                            break;
                        }
                        $arrData = unpack('v', $arrBuf);
                        $intAddr = $arrData[1];
                        if ($intAddr < $ptrRom_a->lngROMSize) 
                        {
                            $byte = $ptrRom_a->ptrROMData[$intAddr];
                            if (fwrite($ptrOutput, $byte) === false) 
                            {
                                break;
                            }
                        }
                    }
                    
                    if ($intI == $lngSlots) 
                    {
                        $blnSuccess = true;
                    }
                    
                    fclose($ptrOutput);
                }
            }
            fclose($ptrInput);
        }
    }
    
    return $blnSuccess;
}

// Test harness
function main() 
{
    $intResult = 1;
    $ptrRom = null;
    $blnDecodeOk = false;
    
    echo "ZOSCII Decoder\n";
    echo "(c) 2026 Cyborg Unicorn Pty Ltd v20260301 - MIT License\n\n";
    
    // Test harness - hardcoded filenames for testing
    $strRomFile = 'rom.bin';
    $strInputFile = 'encoded.enc';
    $strOutputFile = 'decoded.dat';
    
    echo "Test Harness - Using:\n";
    echo "  ROM file: {$strRomFile}\n";
    echo "  Input file: {$strInputFile}\n";
    echo "  Output file: {$strOutputFile}\n\n";
    
    if (file_exists($strRomFile)) 
    {
        $ptrRom = loadRom($strRomFile);
        if ($ptrRom) 
        {
            if (file_exists($strInputFile)) 
            {
                $blnDecodeOk = decodeFile($ptrRom, $strInputFile, $strOutputFile);
                
                if ($blnDecodeOk) 
                {
                    $intResult = 0;
                    echo "Decode successful!\n";
                } 
                else 
                {
                    echo "Decode failed\n";
                }
            } 
            else 
            {
                echo "Input file not found: {$strInputFile}\n";
            }
            
            unloadRom($ptrRom);
        } 
        else 
        {
            echo "Failed to load ROM: {$strRomFile}\n";
        }
    } 
    else 
    {
        echo "ROM file not found: {$strRomFile}\n";
    }
    
    exit($intResult);
}

// Run the test harness
main();
?>